Source/EventFlow/PublishRecovery/PublishVerificator.cs

```diff
@@ -0,0 +1,146 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EventFlow.Aggregates;
+using EventFlow.EventStores;
+
+namespace EventFlow.PublishRecovery
+{
+    public sealed class PublishVerificator : IPublishVerificator
+    {
+        private const int PageSize = 200;
+
+        private readonly IEventPersistence _eventPersistence;
+        private readonly IRecoveryHandlerProcessor _recoveryHandlerProcessor;
+        private readonly IEventJsonSerializer _eventSerializer;
+        private readonly IRecoveryDetector _recoveryDetector;
+        private readonly IReliablePublishPersistence _reliablePublishPersistence;
+
+        public PublishVerificator(IEventPersistence eventPersistence, IRecoveryHandlerProcessor recoveryHandlerProcessor, IEventJsonSerializer eventSerializer, IRecoveryDetector recoveryDetector, IReliablePublishPersistence reliablePublishPersistence)
+        {
+            _eventPersistence = eventPersistence;
+            _recoveryHandlerProcessor = recoveryHandlerProcessor;
+            _eventSerializer = eventSerializer;
+            _recoveryDetector = recoveryDetector;
+            _reliablePublishPersistence = reliablePublishPersistence;
+        }
+
+        public async Task<PublishVerificationResult> VerifyOnceAsync(CancellationToken cancellationToken)
+        {
+            var state = await _reliablePublishPersistence.GetUnverifiedItemsAsync(PageSize, cancellationToken)
+                .ConfigureAwait(false);
+
+            var logItemLookup = state.Items.ToLookup(x => x.AggregateId);
+
+            var page = await _eventPersistence.LoadAllCommittedEvents(state.LastVerifiedPosition, PageSize, cancellationToken)
+                .ConfigureAwait(false);
+
+            var verifyResult = VerifyDomainEvents(page, logItemLookup);
+
+            // Some of not published events can be in flight, so no need recovery them
+            // but we have to check them again on next iteration
+            var eventsForRecovery = GetEventsForRecovery(verifyResult.UnpublishedEvents);
+
+            if (eventsForRecovery.Any())
+            {
+                // Do it inside transaction to recover in single thread
+                // success recovery should put LogItem
+                await _recoveryHandlerProcessor.RecoverAfterUnexpectedShutdownAsync(eventsForRecovery, cancellationToken)
```

Comment:
Isn't this too specific? We don't really know what happened. Maybe just call it "PublishUnconfirmedEvents" or so?

Reply:
Renamed to RecoverUnconfirmedEvents as in recovery strategy we can just skip events (we can decide to do not publish them).